### PR TITLE
Logging: Scope loggers to cell

### LINKED ISSUE #2121: @async run(`sleep 3`) is not async
https://user-images.githubusercontent.com/6933510/169283887-63dda4f6-2dcb-48b2-b0b0-6fba489733ca.mov

### PR DESCRIPTION
**Pull request by Pangoraw**
Alternate implementation based on #2216. This allows to use the same logger in an asynchronous manner. The logs
will be displayed under the right cell automatically because we can keep
the writing task for longer. This fixes #2121.

async_io

### TODO:
 - [x] Logs from macro expansion are dropped when running the cell (https://github.com/fonsp/Pluto.jl/blob/338ca1952485190b315680f0fe70faf2b99adc07/src/evaluation/Run.jl#L164). Find a way to keep them around.

### CODE REVIEW

**Comment by disberd** [comment]
Just checked this out to see if by chance it would also solve #2196 and I noticed that logs generated inside the macro body (so evaluated at _compile_ time) are completely lost, they are not attached to any cell anymore and they do not appear in the julia session that started Pluto either.
Printing to stdout within the macro body is sent to the julia session where Pluto was started, so the complete loss of logs only happens for stuff like `@info`.

**Comment by fonsp** [comment]
@Pangoraw could you update the TODO list? Is the global logger issue the only thing left before merging?

**Comment by Pangoraw** [comment]
> @Pangoraw could you update the TODO list? 

The most problematic is point raised by @disberd where macro expansion logs are dropped when the cell is run, making them impossible to read.

> Is the global logger issue the only thing left before merging?

From my understanding, we can safely remove the global logger since there should be only a few non-trivial way to escape local cell loggers anyway. What do you think?

**Comment by Pangoraw** [comment]
I added log capture during expansion + replay when the cell is finally executed.

**Comment by fonsp** [comment]
Awesome work Paul! I will take a look soon 🎉

File: src/runner/PlutoRunner.jl [hunk 1]
```diff
@@ -2282,8 +2333,31 @@ function with_io_to_logs(f::Function; enabled::Bool=true, loglevel::Logging.LogL
...
+    function send_output()
+        output_str = String(take!(output))
+        if !isempty(output_str)
+            Logging.@logmsg loglevel output_str
+        end
+    end
+    @async begin
+        pipe_reader = Base.pipe_reader(pipe)
+        try
+            while !eof(pipe_reader)
```
Comment (fonsp) [inline src/runner/PlutoRunner.jl:2347]: I expected this short `while` loop to consume 100% CPU, but it doesn't. Nice! Do you know why?

Comment (fonsp) [reply src/runner/PlutoRunner.jl:2347]: Is `readavailable` blocking?

File: src/runner/PlutoRunner.jl [hunk 2]
```diff
@@ -2203,55 +2228,81 @@ const original_stderr = stderr
...
+    stream
+    logger::PlutoCellLogger
+    logs::Vector{Any}
+end
+
+Logging.shouldlog(cl::CaptureLogger, args...) = Logging.shouldlog(cl.logger, args...)
+Logging.min_enabled_level(::CaptureLogger) = min(Logging.Debug, stdout_log_level)
+Logging.catch_exceptions(::CaptureLogger) = Logging.catch_exceptions(cl.logger)
+function Logging.handle_message(pl::CaptureLogger, level, msg, _module, group, id, file, line; kwargs...)
+    Logging.handle_message(pl.logger, level, msg, _module, group, id, file, line; kwargs...)
```
Comment (fonsp) [inline src/runner/PlutoRunner.jl:2252]: what is this line for?

Comment (fonsp) [reply src/runner/PlutoRunner.jl:2252]: It looks like it creates this behaviour. It happens very fast, so I made a video and screenshotted three consecutive frames:

Scherm­afbeelding 2022-11-09 om 13 20 45

Scherm­afbeelding 2022-11-09 om 13 20 59

Scherm­afbeelding 2022-11-09 om 13 21 06

**Review by fonsp — APPROVED** [review]

File: src/runner/PlutoRunner.jl (hunk 1 above)
Comment (Pangoraw) [reply src/runner/PlutoRunner.jl:2357]: From my understanding this is blocking until at least one byte is available (with `wait_readnb(this,1)`):
```julia
function readavailable(this::LibuvStream)
    wait_readnb(this, 1) # unlike the other `read` family of functions, this one doesn't guarantee error reporting
    iolock_begin()
    buf = this.buffer
    @assert buf.seekable == false
    bytes = take!(buf)
    iolock_end()
    return bytes
end
```

**Pull request merged**